### PR TITLE
chore: Update datafusion again

### PR DESCRIPTION
**Pull request by alamb**
# Rationale
* Keep up with changes in DataFusion.
* Test if https://github.com/apache/arrow-datafusion/pull/4122 causes us any problems
* Get fix for https://github.com/influxdata/influxdb_iox/issues/5875 (will test after this is merged)
* Get fix for https://github.com/apache/arrow-datafusion/issues/3968

# Changes:
1. Update DataFusion pin to https://github.com/apache/arrow-datafusion/commit/d2814c960168b45c4a0f5d7bbb72d9f412cb08bd
2. Update APIs for differences

# Required parts
- [x] Update  https://github.com/apache/arrow-datafusion/issues/4169
- [x] Write tests for ParquetChunkList
- [x] Fixed what adding an unecessary sort (very likely https://github.com/apache/arrow-datafusion/pull/4122)
Screenshot 2022-11-10 at 10 23 22 AM

### CODE REVIEW

File: query_tests/cases/in/duplicates_ingester.expected [hunk 1]
```diff
@@ -1,91 +1,95 @@
...
+|               |     ProjectionExec: expr=[time@5 as time, state@4 as state, city@1 as city, min_temp@3 as min_temp, max_temp@2 as max_temp, area@0 as area]                            |
+|               |       RepartitionExec: partitioning=RoundRobinBatch(4)                                                                                                                 |
+|               |         UnionExec                                                                                                                                                      |
+|               |           DeduplicateExec: [state@4 ASC,city@1 ASC,time@5 ASC]                                                                                                         |
+|               |             SortPreservingMergeExec: [state@4 ASC,city@1 ASC,time@5 ASC]                                                                                               |
+|               |               SortExec: [state@4 ASC,city@1 ASC,time@5 ASC]                                                                                                            |
```
Comment (alamb) [inline query_tests/cases/in/duplicates_ingester.expected:17]: This diff doesn't make it easy to see, but the difference is that a new SortExec is added here

The reason this happens is that the new `BasicEnforcement` rule added in https://github.com/apache/arrow-datafusion/pull/4122 by @mingmwang is (correctly) deciding that since the `ParquetExec` declares its output is not sorted, then the output of the `UnionExec` is also therefore not sorted, and thus needs to be resorted prior to the sort preserving merge.

 I think the way to avoid this is to teach DataFusion that the ParquetExec is actually sorted (which is is) and then everything will work out.


As a bonus, I think at least some part of the logic in the plan construction of IOx (that adds SortExec's in to sort the data) could potentially be removed as it is now covered by the DataFusion optimizer. I think this is very much in line with  @crepererum 's ideas from https://github.com/influxdata/influxdb_iox/issues/6098

**Review by crepererum — APPROVED** [review]

File: query_tests/cases/in/duplicates_ingester.expected (hunk 1 above)
Comment (alamb) [reply query_tests/cases/in/duplicates_ingester.expected:17]: Filed upstream issue: https://github.com/apache/arrow-datafusion/issues/4169

File: iox_query/src/provider/physical.rs [hunk 2]
```diff
@@ -86,31 +137,42 @@ pub fn chunks_to_physical_nodes(
...
                 partition_values: vec![],
                 range: None,
                 extensions: None,
             }),
             target_partitions,
         );
+
+        // Tell datafusion about the sort key, if any
+        let file_schema = iox_schema.as_arrow();
+        let output_ordering =
```
Comment (alamb) [inline iox_query/src/provider/physical.rs:159]: Creating this output ordering is important to tell  DataFusion the data is already sorted so that it doesn't put another sort in our plan

File: querier/src/namespace/query_access.rs [hunk 3]
```diff
@@ -508,24 +508,24 @@ mod tests {
...
+                "|               |     ProjectionExec: expr=[host@0 as host, perc@1 as perc, time@2 as time]                                                                                                       |",
+                "|               |       UnionExec                                                                                                                                                                 |",
+                "|               |         CoalesceBatchesExec: target_batch_size=4096                                                                                                                             |",
+                "|               |           FilterExec: time@2 < 1 OR time@2 > 13 OR NOT host@0 = CAST(d AS Dictionary(Int32, Utf8))                                                                              |",
+                "|               |             ParquetExec: limit=None, partitions=[1/2/1/4/<uuid>.parquet], output_ordering=[host@0 ASC, time@2 ASC], projection=[host, perc, time] |",
```
Comment (alamb) [inline querier/src/namespace/query_access.rs:523]: the difference in these plans is that the output ordering is now specified

**Comment by alamb** [comment]
@crepererum  FYI this PR has changed substantially since you last reviewed it

File: iox_query/src/provider/physical.rs [hunk 4]
```diff
@@ -15,12 +18,58 @@ use datafusion::{
...
+    if let (Some(existing_sort_key), Some(chunk_sort_key)) = (existing_sort_key, chunk_sort_key) {
+        let combined_sort_key = SortKey::try_merge_key(&existing_sort_key, chunk_sort_key);
+
+        match combined_sort_key {
+            Some(combined_sort_key) if combined_sort_key == &existing_sort_key => {
+                Some(existing_sort_key)
+            }
+            Some(combined_sort_key) => Some(combined_sort_key.clone()),
+            None => None,
+        }
```
Comment (NGA-TRAN) [inline iox_query/src/provider/physical.rs:66]: So the whole purpose of this match is to not clone the key when possible?

File: iox_query/src/provider/physical.rs [hunk 5]
```diff
@@ -63,12 +112,14 @@ pub fn chunks_to_physical_nodes(
                 let url_str = parquet_input.object_store_url.as_str().to_owned();
                 match parquet_chunks.entry(url_str) {
                     Entry::Occupied(mut o) => {
-                        o.get_mut().1.push(parquet_input.object_meta);
+                        o.get_mut()
+                            .add_parquet_file(chunk.as_ref(), parquet_input.object_meta);
```
Comment (NGA-TRAN) [inline iox_query/src/provider/physical.rs:116]: I asked myself why the `combine_sort_key` returns `None` when one of the input is none. But when I come to this line I see it makes sense 👍

File: iox_query/src/provider/physical.rs [hunk 6]
```diff
@@ -156,4 +220,48 @@ mod tests {
...
+
+        assert_eq!(
+            combine_sort_key(Some(skey_t1_t2.clone()), Some(&skey_t1)),
+            Some(skey_t1_t2)
+        );
+
+        assert_eq!(
+            combine_sort_key(Some(skey_t2_t1.clone()), Some(&skey_t1)),
+            Some(skey_t2_t1)
+        );
```
Comment (NGA-TRAN) [inline iox_query/src/provider/physical.rs:265]: You may want to add a test to combine `skey_t1_t2` with `skey_t2_t1` and I think it will return None, right?

File: parquet_file/src/storage.rs [hunk 7]
```diff
@@ -123,6 +123,8 @@ impl ParquetExecInput {
             table_partition_cols: vec![],
             // TODO avoid this `copied_config` when config_options are directly available on context
             config_options: session_ctx.copied_config().config_options(),
+            // This is not correct for production: they are ordered
+            output_ordering: None,
```
Comment (NGA-TRAN) [inline parquet_file/src/storage.rs:127]: I am not sure I understand the comment

File: query_tests/cases/in/duplicates_parquet.expected [hunk 8]
```diff
@@ -91,8 +91,8 @@
...
...l_time{filename=1/1/1/1/00000000-0000-0000-0000-000000000002.parquet}=1.234ms, page_index_eval_time{filename=1/1/1/1/00000000-0000-0000-0000-000000000003.parquet}=1.234ms, page_index_rows_filtered{filename=1/1/1/1/00000000-0000-0000-0000-000000000002.parquet}=0, page_index_rows_filtered{filename=1/1/1/1/00000000-0000-0000-0000-000000000003.parquet}=0, predicate_evaluation_errors{filename=1/1/1/1/00000000-0000-0000-0000-000000000002.parquet}=0, predicate_evaluation_errors{filename=1/1/1/1/00000000-0000-0000-0000-000000000003.parquet}=0, pushdown_eval_time{filename=1/1/1/1/00000000-0000-0000-0000-000000000002.parquet}=1.234ms, pushdown_eval_time{filename=1/1/1/1/00000000-0000-0000-0000-000000000003.parquet}=1.234ms, pushdown_rows_filtered{filename=1/1/1/1/00000000-0000-0000-0000-000000000002.parquet}=2, pushdown_rows_filtered{filename=1/1/1/1/00000000-0000-0000-0000-000000000003.parquet}=3, row_groups_pruned{filename=1/1/1/1/00000000-0000-0000-0000-000000000002.parquet}=0, row_groups_pruned{filename=1/1/1/1/00000000-0000-0000-0000-000000000003.parquet}=0, spill_count=0, spilled_bytes=0, time_elapsed_opening=1.234ms, time_elapsed_processing=1.234ms, time_elapsed_scanning=1.234ms]    |
```
Comment (NGA-TRAN) [inline query_tests/cases/in/duplicates_parquet.expected:96]: Even though there are no highlight here, the differences are the output-ordering 👍

**Review by NGA-TRAN — APPROVED** [review]
So nice that we can see the output ordering in the plan now and that we upgrade DF but keep our query plan optimal. Thanks @alamb
Those query plans are worth keeping

File: iox_query/src/exec/non_null_checker.rs [hunk 9]
```diff
@@ -218,8 +218,8 @@ impl ExecutionPlan for NonNullCheckerExec {
         None
     }
 
-    fn required_child_distribution(&self) -> Distribution {
-        Distribution::UnspecifiedDistribution
+    fn required_input_distribution(&self) -> Vec<Distribution> {
```
Comment (crepererum) [inline iox_query/src/exec/non_null_checker.rs:221]: Is the distribution "per input node" now instead of "for all inputs"?

File: parquet_file/src/storage.rs [hunk 10]
```diff
@@ -123,6 +123,8 @@ impl ParquetExecInput {
             table_partition_cols: vec![],
             // TODO avoid this `copied_config` when config_options are directly available on context
             config_options: session_ctx.copied_config().config_options(),
+            // This is not correct for production: they are ordered
```
Comment (crepererum) [inline parquet_file/src/storage.rs:126]:
```suggestion
            // Parquet files ARE actually sorted but we don't care here since we just construct a `collect` plan.
```

File: querier/src/ingester/flight_client.rs [hunk 11]
```diff
@@ -224,17 +224,19 @@ mod tests {
         // see https://github.com/influxdata/influxdb_iox/issues/5974
```
[the remark this answers is not in the document]
Comment (crepererum) [reply querier/src/ingester/flight_client.rs:224]: No, we still need this test because there might be a different expression construct hitting this limit. I can construct one for you (using some ugly nested case statements for example).

File: iox_query/src/provider/physical.rs [hunk 12]
```diff
@@ -15,12 +18,58 @@ use datafusion::{
...
+
+/// Combines the existing sort key with the sort key of the chunk, returning the new sort key.
+fn combine_sort_key(
+    existing_sort_key: Option<SortKey>,
+    chunk_sort_key: Option<&SortKey>,
+) -> Option<SortKey> {
+    if let (Some(existing_sort_key), Some(chunk_sort_key)) = (existing_sort_key, chunk_sort_key) {
+        let combined_sort_key = SortKey::try_merge_key(&existing_sort_key, chunk_sort_key);
+
+        match combined_sort_key {
```
Comment (crepererum) [inline iox_query/src/provider/physical.rs:60]: `combined_sort_key.map(|k| if ... {...} else {...})` is the more Rusty way, but that's optional

File: iox_query/src/provider/physical.rs [hunk 13]
```diff
@@ -15,12 +18,58 @@ use datafusion::{
...
+            Some(combined_sort_key) if combined_sort_key == &existing_sort_key => {
+                Some(existing_sort_key)
+            }
+            Some(combined_sort_key) => Some(combined_sort_key.clone()),
+            None => None,
+        }
+    } else {
+        // no existing sort key means the data wasn't consistently sorted so leave it alone
```
Comment (crepererum) [inline iox_query/src/provider/physical.rs:68]: This will happen in prod. The issue is that we have chunks for different partitions here (I think) and partitions have different sort orders. You may chose to NOT address this now, since this requires changes to the chunk grouping, but then we at least need to track this in a somewhat-urgent (= next week) follow-up ticket.

**Review by crepererum — CHANGES_REQUESTED** [review]
Needs some more work regarding the overflow test (see comment) and the parquet chunk grouping due to inter-partition sort key inconsistencies.

File: iox_query/src/exec/non_null_checker.rs (hunk 9 above)
Comment (alamb) [reply iox_query/src/exec/non_null_checker.rs:221]: Yes, exactly, it is per-child, to handle cases like joins where the children may / may not be partitioned correctly to perform a parallel equijoin without redistributing the inputs

File: iox_query/src/provider/physical.rs (hunk 4 above)
Comment (alamb) [reply iox_query/src/provider/physical.rs:73]: yes, that is correct -- I expect the sort key will commonly be the same so avoiding the cloning will be substantial. I will make this clear in comments.

File: iox_query/src/provider/physical.rs (hunk 13 above)
Comment (alamb) [reply iox_query/src/provider/physical.rs:75]: It does sound somewhat concerning, though I don't fully follow what the concern is -- if we are merging chunks from different partitions with different sort orders together that will result in wrong results.

File: iox_query/src/provider/physical.rs (hunk 12 above)
Comment (alamb) [reply iox_query/src/provider/physical.rs:67]: I tried doing this and the borrow checker was not happy:

```
error[E0505]: cannot move out of `existing_sort_key` because it is borrowed
  --> iox_query/src/provider/physical.rs:65:31
   |
63 |         let combined_sort_key = SortKey::try_merge_key(&existing_sort_key, chunk_sort_key);
   |                                                        ------------------ borrow of `existing_sort_key` occurs here
64 |
65 |         combined_sort_key.map(|combined_sort_key| {
   |                           --- ^^^^^^^^^^^^^^^^^^^ move out of `existing_sort_key` occurs here
   |                           |
   |                           borrow later used by call
...
69 |                 existing_sort_key
   |                 ----------------- move occurs due to use in closure
```

File: iox_query/src/provider/physical.rs (hunk 5 above)
Comment (alamb) [reply iox_query/src/provider/physical.rs:123]: I added some comments to `combine_sort_key` to try and make this clearer

File: iox_query/src/provider/physical.rs (hunk 6 above)
Comment (alamb) [reply iox_query/src/provider/physical.rs:272]: Agreed. Added in e5c1365cd

File: parquet_file/src/storage.rs (hunk 7 above)
Comment (alamb) [reply parquet_file/src/storage.rs:127]: I think @crepererum 's rewording is much better -- does that make more sense?

```rust
 // Parquet files ARE actually sorted but we don't care here since we just construct a `collect` plan.
```

?

File: querier/src/ingester/flight_client.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (alamb) [reply querier/src/ingester/flight_client.rs:224]: Ah, I see -- I will see if I can construct such a thing

File: querier/src/ingester/flight_client.rs [hunk 14]
```diff
@@ -231,28 +252,54 @@ mod tests {
...
-                    predicate: Some(predicate),
-                };
+                // build a deeply recursive nested expression:
+                //
+                // CASE
+                //  WHEN TRUE
+                //  THEN (WHEN ...)
+                // ELSE FALSE
+                //
+                let expr = (0..n).fold(lit(false), |expr, _|{
```
Comment (alamb) [inline querier/src/ingester/flight_client.rs:263]: I have made another deeply nested expression as suggested by @crepererum  in addition to a second test that verifies deeply nested `AND`s now work

File: querier/src/ingester/flight_client.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (alamb) [reply querier/src/ingester/flight_client.rs]: done

File: iox_query/src/provider/physical.rs (hunk 13 above)
Comment (crepererum) [reply iox_query/src/provider/physical.rs:75]: Ah, because this is only relevant for de-dup, right? Well, then let's leave it. If it is a performance issue in some way (because we insert merges that shouldn't be there), we can address this later.

**Review by crepererum — APPROVED** [review]

File: iox_query/src/provider/physical.rs (hunk 13 above)
Comment (alamb) [reply iox_query/src/provider/physical.rs:75]: yeah -- I hope once we have completed the predicate pushdown work (we are very close) and "run on any data size" then we can move on to profiling and improving performance for large numbers of parquet files
